[FEATURE] Ajout d'une page pour la création de configurations de score et de niveau par compétence (PIX-11639).

alexandrecoin opened the pull request:
## :unicorn: Problème

Les données de scoring actuelles ne sont plus valables.
Un nouveau fichier JSON a été transmis par l’équipe Data avec les nouvelles mailles. 

## :robot: Proposition

Puisque l'ensemble des configurations (score de certification et niveaux par compétence) est utilisée de façon conjointe, nous proposons de réunir ces deux configurations dans un seul et même objet.

## :rainbow: Remarques

- La route créée pour l'occasion côté front ne contient pas de tests automatisés tout simplement car elle n'a pas vocation a rester et nous permet simplement de pouvoir laisser les équipes produit/métier travailler en autonomie sur l'ajout d'une configuration (peu probable dans un futur proche) sans avoir à passer par des requêtes curl contraignantes et sources d'erreur de frappe.

## :100: Pour tester

Dans pix-admin, se connecter en tant que super admin : 
- Aller sur la page `/certifications/configuration`
- Ajouter les deux configurations suivantes : 

`Niveau par compétence`
```
[
  {
    "competence": "1.1",
    "values": [
      {
        "bounds": {
          "max": -2.123,
          "min": -6.789
        },
        "competenceLevel": 0
      },
      {
        "bounds": {
          "max": -1.234,
          "min": -5.678
        },
        "competenceLevel": 1
      }
    ]
  }
]
```


`Score de certification`
```
[
    {"meshLevel": 0, "bounds": {"min": -8, "max": -4}},
    {"meshLevel": 1, "bounds": {"min": -4, "max": 0}},
    {"meshLevel": 2, "bounds": {"min": 0, "max": 2}},
    {"meshLevel": 3, "bounds": {"min": 2, "max": 4}},
    {"meshLevel": 4, "bounds": {"min": 4, "max": 5}},
    {"meshLevel": 5, "bounds": {"min": 5, "max": 6}},
    {"meshLevel": 6, "bounds": {"min": 6, "max": 7}},
    {"meshLevel": 7, "bounds": {"min": 7, "max": 8}}
]
```

Pour aller plus loin dans le test :
- Créer et passer une certification V3 (`certifv3@example.net` et `certifiable-contenu-user@example.net`)
- Terminer le test
- Vérifier que le score du candidat a bien été généré.

cberthou committed:
:sparkles: admin: add ui to create scoring configurations

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr8414.review.pix.fr)
  * [App (.org)](https://app-pr8414.review.pix.org)
  * [Orga (.fr)](https://orga-pr8414.review.pix.fr)
  * [Orga (.org)](https://orga-pr8414.review.pix.org)
  * [Certif (.fr)](https://certif-pr8414.review.pix.fr)
  * [Certif (.org)](https://certif-pr8414.review.pix.org)
  * [1D](https://1d-pr8414.review.pix.fr)
  * [Admin](https://admin-pr8414.review.pix.fr)
  * [API](https://api-pr8414.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr8414.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr8414/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr8414/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr8414/environment)

Commit:
:recycle: api: rename competence-for-scoring-repository

Renaming to scoring-configuration-repository as it will contain
both general scoring and competence level related methods

Commit:
:card_file_box: api: create scoring-configurations table

Commit:
:sparkles: api: add scoring configuration DB builder

Commit:
:seedling: api: add scoring configuration seed

cberthou committed:
:recycle: api: extracted v3 certification logic into a specific domain object

Commit:
:recycle: api: rename listByLocale to getLatestByDateAndLocale

Commit:
:sparkles: api: add save method for scoring configuration

Commit:
:recycle: api: rename save method in scoring-configuration-repository

This renaming allows us to add the save mmethod for certification scoring configuration
without confusion

Commit:
:recycle: api: rename competences-for-scoring route and contoller to scoring-configuration

This renaming will allow us to gather both certification scoring and competences scoring
related routes

Commit:
:sparkles: api: create route for adding a certification scoring config

Commit:
:sparkles: api: use real certification scoring conf instead of hard coded one

mcampourcy commented:
Test func KO, erreur 500 à la fin de la certif

Commit:
:sparkles: api: use minimum and maximum mesh level values instead of hardcoded values

Using hardcoded minimum and maximum values forces us to change the minimum and maximum values
provided by the data team on configuration importation as the candidate level must be part of an interval
in order to be scored

mcampourcy commented:
Tech review OK, func review OK

<img width="772" alt="Capture d’écran 2024-03-19 à 14 14 03" src="https://github.com/1024pix/pix/assets/11388201/3086be12-e10b-4e18-8750-9a18ed16a50b">

aceol commented on api/db/migrations/20240315094742_create-certification-scoring-configurations-table.js:
```diff
@@ -0,0 +1,15 @@
+  await knex.schema.createTable(TABLE_NAME, function (table) {
+    table.increments('id').primary();
+    table.jsonb('configuration').notNullable();
+    table.dateTime('createdAt').notNullable().defaultTo(knex.fn.now());
```
pas besoin du `updatedAt`? Ca semblerait pertinent ici

AndreiaPena replied:
quid de l'id de la personne qui aurait fait l'action ?

alexandrecoin replied:
@aceol On update pas les config mais on en ajoute uniquement afin de pouvoir se baser sur la configuration exacte qui a été utilisée par le candidat en cas de rescoring par exemple.

alexandrecoin replied:
@AndreiaPena Effectivement, on va l'ajouter !

aceol replied:
Ah mais on ne fait que rajouter, donc pas besoin...

aceol commented on api/lib/domain/events/handle-certification-rescoring.js:
```diff
@@ -123,15 +123,15 @@ async function _handleV3Certification({
   });
 
-  const competencesForScoring = await scoringConfigurationRepository.listByLocale({ locale });
+  const v3CertificationScoring = await scoringConfigurationRepository.listByLocale({ locale });
```
Je trouve que le v3 à la fin simplifie la lecture
```suggestion
  const certificationScoringV3 = await scoringConfigurationRepository.listByLocale({ locale });
```

aceol replied:
Commentaire pour la classe associé :D

aceol commented on api/tests/certification/scoring/unit/domain/models/V3CertificationScoring_test.js:
```diff
@@ -0,0 +1,77 @@
+    });
+  });
+
+  describe('.fromConfigurations', function () {
```
Et-ce que le . c'est parce que static, ou une typo?
```suggestion
  describe('#fromConfigurations', function () {
```

alexandrecoin replied:
Typo.

AndreiaPena commented on admin/app/templates/authenticated/certifications/configuration.hbs:
```diff
@@ -0,0 +1,19 @@
+<form {{on "submit" this.onCompetenceScoringConfigurationSubmit}}>
+  <label>Ajout de la configuration de niveau par compétence (JSON) :</label>
+  <PixTextarea
+    id="competence-configuration"
+    label="Competence scoring configuration"
+    {{on "change" this.onCompetenceScoringConfigurationChange}}
+  />
+  <PixButton type="submit">Envoyer</PixButton>
+</form>
+
+<form {{on "submit" this.onCertificationScoringConfigurationSubmit}}>
+  <label>Ajout de la configuration de score de certification (JSON) :</label>
+  <PixTextarea
+    id="certification-scoring-configuration"
+    label="certification scoring configuration"
+    {{on "change" this.onCertificationScoringConfigurationChange}}
+  />
+  <PixButton type="submit">Envoyer</PixButton>
+</form>
```
**Suggestion** : Histoire de pimper un peu avec une section blanche et valider l'a11y avec le bon usage du composant `PixTextarea`

```suggestion
<section class="page-section">
    <form {{on "submit" this.onCompetenceScoringConfigurationSubmit}}>
    <PixTextarea
      @id="competence-configuration"
      @label="Ajout de la configuration de niveau par compétence (JSON) :"
      {{on "change" this.onCompetenceScoringConfigurationChange}}
    />
    <PixButton @type="submit">Envoyer</PixButton>
  </form>

  <form {{on "submit" this.onCertificationScoringConfigurationSubmit}}>
    <PixTextarea
      @id="certification-scoring-configuration"
      @label="Ajout de la configuration de score de certification (JSON) :"
      {{on "change" this.onCertificationScoringConfigurationChange}}
    />
    <PixButton @type="submit">Envoyer</PixButton>
</form>
</section>
```

AndreiaPena replied:
Je propose aussi d'ajouter un fichier `configuration.scss` avec ceci : 

```
.certification-configuration-form {

  &:first-child {
    padding-bottom: var(--pix-spacing-8x);
  }

  textarea {
    min-height: 300px;
    max-width: 75%;
  }
}
```
La classe serait placée sur les 2 form. Ça permettra de voir un peu le contenu qu'on y met.

<img width="1318" alt="Capture d’écran 2024-03-20 à 10 39 40" src="https://github.com/1024pix/pix/assets/58915422/1b37e1d0-40d2-4a3d-8eae-a3763db41d16">

AndreiaPena commented on admin/app/controllers/authenticated/certifications/configuration.js:
```diff
@@ -0,0 +1,46 @@
+      await adapter.updateCompetenceScoringConfiguration(this.competenceScoringConfiguration);
+      this.competenceScoringConfiguration = '';
+      this.notifications.success('Configuration mise à jour');
+    } catch (e) {
+      this.notifications.error(`La config n'a pu être ajoutée: ${e}`);
```
**Suggestion** : Pas très clair le message d'erreur 😅 

<img width="415" alt="Capture d’écran 2024-03-20 à 10 15 27" src="https://github.com/1024pix/pix/assets/58915422/a6c07a22-736a-4b93-9bda-01ca82ced57b">


```suggestion
    } catch (error) {
      this.notifications.error(`La configuration n'a pas pu être ajoutée: ${{error.errors[0].detail}}`);
```

Au moins on a le détail de l'erreur seulement

<img width="415" alt="Capture d’écran 2024-03-20 à 10 23 22" src="https://github.com/1024pix/pix/assets/58915422/8cebd5d4-d406-467c-8140-d497f8bbea82">

AndreiaPena replied:
Mais à choisir je préfère ne pas dépendre du back si on a pas géré les erreurs avec des codes etc.. et juste mettre coté front `La configuration n'a pas pu être ajoutée`

AndreiaPena commented on admin/app/controllers/authenticated/certifications/configuration.js:
```diff
@@ -0,0 +1,46 @@
+      await adapter.updateCertificationScoringConfiguration(this.certificationScoringConfiguration);
+      this.certificationScoringConfiguration = '';
+      this.notifications.success('Configuration mise à jour');
+    } catch (e) {
+      this.notifications.error(`La config n'a pu être ajoutée: ${e}`);
```
```suggestion
    } catch (error) {
      this.notifications.error(`La configuration n'a pas pu être ajoutée: ${{error.errors[0].detail}}`);
```

<img width="415" alt="Capture d’écran 2024-03-20 à 10 30 24" src="https://github.com/1024pix/pix/assets/58915422/bdf2bf5e-dfe9-4e45-b277-c1ad05727b91">

AndreiaPena commented on api/db/seeds/data/team-certification/data-builder.js:
```diff
@@ -60,6 +61,7 @@ async function teamCertificationDataBuilder({ databaseBuilder }) {
   await _createComplementaryCertificationCampaign({ databaseBuilder });
   _createV3CertificationConfiguration({ databaseBuilder });
   createCompetenceScoringConfiguration({ databaseBuilder });
+  createScoringConfiguration({ databaseBuilder });
```
**chipotage** : 
```suggestion
  _createCompetenceScoringConfiguration({ databaseBuilder });
  _createScoringConfiguration({ databaseBuilder });
```

alexandrecoin replied:
Ces méthodes sont importées de fichiers externes car très volumineuses à la lecture, pas vraiment privées donc.

AndreiaPena commented on admin/app/router.js:
```diff
@@ -83,6 +83,7 @@ Router.map(function () {
     });
 
     this.route('certifications', function () {
+      this.route('configuration');
```
**Remarque** : il manque des tests coté front.

AndreiaPena commented on api/src/certification/scoring/application/scoring-configuration-route.js:
```diff
@@ -0,0 +1,82 @@
+  server.route([
+    {
+      method: 'POST',
+      path: '/api/competence-for-scoring-configuration',
```
```suggestion
      path: '/api/admin/competence-for-scoring-configuration',
```

AndreiaPena commented on api/src/certification/scoring/application/scoring-configuration-route.js:
```diff
@@ -0,0 +1,82 @@
+            .required(),
+        },
+        handler: scoringConfigurationController.saveCompetenceForScoringConfiguration,
+        tags: ['api', 'scoring-configuration'],
```
```suggestion
        tags: ['api', 'admin', 'scoring-configuration'],
```

AndreiaPena commented on api/src/certification/scoring/application/scoring-configuration-route.js:
```diff
@@ -0,0 +1,82 @@
+  server.route([
+    {
+      method: 'POST',
+      path: '/api/certification-scoring-configuration',
```
```suggestion
      path: '/api/admin/certification-scoring-configuration',
```

AndreiaPena commented on api/src/certification/scoring/application/scoring-configuration-route.js:
```diff
@@ -0,0 +1,82 @@
+            .required(),
+        },
+        handler: scoringConfigurationController.saveCertificationScoringConfiguration,
+        tags: ['api', 'scoring-configuration'],
```
```suggestion
        tags: ['api', 'admin', 'scoring-configuration'],
```

AndreiaPena commented on api/tests/certification/scoring/acceptance/application/scoring-configuration-route_test.js:
```diff
@@ -139,4 +139,121 @@ describe('Acceptance | Application | competence-for-scoring-configuration-route'
+  describe('POST /api/certification-scoring-configuration', function () {
+    describe('when called without being authenticated', function () {
+      it('should return a 401', async function () {
+        // given
+        const options = {
+          method: 'POST',
+          url: '/api/certification-scoring-configuration',
+        };
+        // when
+        const response = await server.inject(options);
+        // then
+        expect(response.statusCode).to.equal(401);
```
```suggestion
        // given
        const options = {
          method: 'POST',
          url: '/api/certification-scoring-configuration',
        };
        
        // when
        const response = await server.inject(options);
        
        // then
        expect(response.statusCode).to.equal(401);
```

aceol commented on api/src/certification/scoring/infrastructure/repositories/scoring-configuration-repository.js:
```diff
@@ -3,17 +3,19 @@ import * as areaRepository from '../../../../shared/infrastructure/repositories/
 import { V3CertificationScoring } from '../../domain/models/V3CertificationScoring.js';
 
-export const listByLocale = async ({ locale }) => {
+export const getMostRecentByDateAndLocale = async ({ locale, date }) => {
```
Est-ce que c'est pareil?
```suggestion
export const getLatestByDateAndLocale = async ({ locale, date }) => {
```

alexandrecoin replied:
It is indeed !

AndreiaPena commented on api/tests/certification/scoring/acceptance/application/scoring-configuration-route_test.js:
```diff
@@ -139,4 +139,121 @@ describe('Acceptance | Application | competence-for-scoring-configuration-route'
       });
```
☝🏻 le test du dessus parle de 204 alors qu'on a une 201

AndreiaPena commented on api/tests/certification/scoring/integration/application/scoring-configuration-controller_test.js:
```diff
@@ -0,0 +1,47 @@
+
+      const response = await scoringConfigurationController.saveCertificationScoringConfiguration(request, hFake);
+
+      expect(response.statusCode).to.equal(201);
```
```suggestion
    // given
      sinon.stub(usecases, 'saveCompetenceForScoringConfiguration');

      const request = {
        payload: {
          data: {
            warmUpLength: 12,
          },
        },
      };

// when
      const response = await scoringConfigurationController.saveCompetenceForScoringConfiguration(request, hFake);

// then
      expect(response.statusCode).to.equal(201);
      expect(usecases.saveCompetenceForScoringConfiguration).to.have.been.calledWith({
        data: request.payload,
      });
    });
  });

  describe('#saveCertificationScoringConfiguration', function () {
    it('should save the certification scoring configuration', async function () {
    // given
      sinon.stub(usecases, 'saveCertificationScoringConfiguration');

      const request = {
        payload: {
          data: {
            someData: 10,
          },
        },
      };
      
// when
      const response = await scoringConfigurationController.saveCertificationScoringConfiguration(request, hFake);

// then
      expect(response.statusCode).to.equal(201);
```

aceol commented on api/tests/certification/scoring/integration/infrastructure/repositories/scoring-configuration-repository_test.js:
```diff
@@ -131,4 +132,19 @@ describe('Unit | Repository | scoring-configuration-repository', function () {
+      await saveCertificationScoringConfiguration(data);
+
+      // then
+      const configurations = await knex('certification-scoring-configurations');
+      expect(configurations.length).to.equal(1);
+      expect(configurations[0].configuration).to.deep.equal({ some: 'data' });
```
Est-ce qu'on ne devrait pas verifier qu'on a bien une date de creation enregistré?

alexandrecoin replied:
Pourquoi pas, mais c'est Postgres qui gère donc on devrait pas avoir à s'en inquiéter.

aceol commented on api/src/certification/scoring/domain/usecases/save-competence-for-scoring-configuration.js:
```diff
@@ -7,7 +7,7 @@
  */
 const saveCompetenceForScoringConfiguration = async ({ data, scoringConfigurationRepository }) => {
-  await scoringConfigurationRepository.save(data);
+  await scoringConfigurationRepository.saveCompetenceForScoringConfiguration(data);
```
Je ne comprends pas le besoin de renommer

alexandrecoin replied:
On enregistre deux configurations différentes dans le même repo.

AndreiaPena commented on api/tests/certification/scoring/integration/infrastructure/repositories/scoring-configuration-repository_test.js:
```diff
@@ -0,0 +1,150 @@
+    });
+  });
+
+  describe('#save', function () {
```
```suggestion
  describe('#saveCompetenceForScoringConfiguration', function () {
```

aceol commented on api/src/certification/scoring/domain/usecases/save-certification-scoring-configuration.js:
```diff
@@ -0,0 +1,13 @@
+ * @param {ScoringConfigurationRepository} params.scoringConfigurationRepository
+ */
+const saveCertificationScoringConfiguration = async ({ data, scoringConfigurationRepository }) => {
+  await scoringConfigurationRepository.saveCertificationScoringConfiguration(data);
```
```suggestion
  return scoringConfigurationRepository.saveCertificationScoringConfiguration(data);
```

aceol commented on api/src/certification/scoring/infrastructure/repositories/scoring-configuration-repository.js:
```diff
@@ -27,7 +27,7 @@ export const getMostRecentByDateAndLocale = async ({ locale, date }) => {
 };
 
-export const save = async (configuration) => {
+export const saveCompetenceForScoringConfiguration = async (configuration) => {
```
On ne souhaite plus creer de methode de repo sans destructuration. Cela permet de pouvoir injecter des api externes si besoin
```suggestion
export const saveCompetenceForScoringConfiguration = async ({ configuration }) => {
```

aceol commented on api/src/certification/scoring/domain/models/CertificationAssessmentScoreV3.js:
```diff
@@ -3,48 +3,8 @@ import { status as CertificationStatus } from '../../../../shared/domain/models/
-/*
-Score should not be totally linear. It should be piecewise linear, so we define
-here the different intervals. See documentation here :
-https://1024pix.atlassian.net/wiki/spaces/DD/pages/3835133953/Vulgarisation+score+2023
```
On ne souhaite pas de lien vers de la doc privé sur le repo open-source. Est-ce que ca a ete discuté?

alexandrecoin replied:
Effectivement, ça n'a pas été discuté. Erreur de notre part.

aceol commented on api/src/certification/scoring/domain/models/CertificationAssessmentScoreV3.js:
```diff
@@ -77,12 +74,14 @@ const _computeScore = ({
 }) => {
   let normalizedEstimatedLevel = estimatedLevel;
+  const minimumEstimatedLevel = certificationScoringIntervals[0].bounds.min;
+  const maximumEstimatedLevel = certificationScoringIntervals[certificationScoringIntervals.length - 1].bounds.max;
```
La beauté du js:
```suggestion
  const maximumEstimatedLevel = certificationScoringIntervals.at(-1).bounds.max;
```

alexandrecoin replied:
Yes, je suis de l'ancienne école encore :D

aceol commented on api/src/certification/scoring/domain/models/CertificationAssessmentScoreV3.js:
```diff
@@ -77,12 +74,14 @@ const _computeScore = ({
 
-  if (normalizedEstimatedLevel < MINIMUM_ESTIMATED_LEVEL) {
-    normalizedEstimatedLevel = MINIMUM_ESTIMATED_LEVEL;
+  if (normalizedEstimatedLevel < minimumEstimatedLevel) {
```
Pas certain qu'il soit judicieux d'utiliser une lib externe mais lodash fait ca:

`_.clamp(maximumEstimatedLevel, minimumEstimatedLevel, normalizedEstimatedLevel);`

alexandrecoin replied:
Je trouve pas ça super lisible perso.

aceol commented on admin/app/adapters/scoring-configuration.js:
```diff
@@ -0,0 +1,8 @@
+export default class ScoringConfigurationAdapter extends ApplicationAdapter {
+  updateCompetenceScoringConfiguration(newConfiguration) {
+    const url = `${this.host}/${this.namespace}/competence-for-scoring-configuration`;
+    return this.ajax(url, 'POST', { data: newConfiguration });
```
On se perde un peu avec les 
`competence scoring configuration`
`scoring configuration`
`competence for scoring configuration` 
`configuration`
L'harmonization permettrait d'eviter l'adapter!?

alexandrecoin replied:
La page va pas rester comme dit dans les remarques de la PR, on a fait au plus rapide sans abstraction.

aceol commented:
> Puisque l'ensemble des configurations (score de certification et niveaux par compétence) est utilisée de façon conjointe, nous proposons de réunir ces deux configurations dans un seul et même objet.

En fait on a deux tables et 2 objets distinct. Je suis un peu confus

Commit:
:recycle: api: use object instead of regular params in scoring repository

Commit:
:recycle: api: alter route for saving competence scoring config

Commit:
:white_check_mark: api: add missing given-when-then

Commit:
:sparkles: api: update usecase and repository with missing userId property

Commit:
:white_check_mark: api: update acceptance tests to meet new createdByUserId column requirements
